### ISSUE TITLE
Warn users when cross entropy is called after softmax

### ISSUE THREAD

**Issue by ab-10** [pytorch/pytorch#55902]
Making an issue to collect some feedback on a potential feature. If this sounds good, I'm happy to implement it.

## 🚀 Feature
When `torch.nn.CrossEntropy_loss` or `F.cross_entropy` is called on a tensor, where the `grad_fn` is `SoftmaxBackward` display a warning.

## Motivation

A common error is to call SoftMax prior to performing cross entropy loss, I believe users should be warned about the wasted computation. Because doing so is far simpler on tensors with grads, I suggest that the warning is limited only to such a scenario.

**Comment by mruberry** [comment]
Thank you for this suggestion, @ab-10. It's an interesting idea.

**Comment by jbschlosser** [comment]
I'm a bit worried about false negatives. If people realize that this warning exists and start to depend on it, it may be confusing to have cases where it should be thrown but isn't. Examples:
* Tensors that don't require grad
* Any operation that changes the `grad_fn` (e.g. view operations) in between softmax and cross entropy

**Comment by ab-10** [comment]
Thank you @jbschlosser for the edge cases, I didn't consider the (real) possibility that users might interpret the absence of a warning as validation. As far as I can tell this limitations cannot be avoided without tracking additional property (e.g. `is_softmaxed`) for every tensor and that seems like an overkill.

Do you think the suggestion would bring more benefit than confusion given these limitations?

**Comment by jbschlosser** [comment]
> Do you think the suggestion would bring more benefit than confusion given these limitations?

I personally think the inconsistency might bring more harm than good, but I'm open to being convinced otherwise. I agree it's overkill to do the full tracking, so I'd lean towards no warnings rather than partial warnings.

**Comment by sachinbareth** [comment]
softmax cross-entropy loss without encountering the error.
